threading/smartptrs.nim

```diff
@@ -0,0 +1,158 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2021 Nim contributors
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+
+## C++11 like smart pointers. They always use the shared allocator.
```

Comment:
I'm still a bit confused why this is called "C++ like smart pointers" - they're nothing like their C++ counterpart - in particular, the C++ counterpart is used to manage *a pointer*  - this library manages a memory buffer that it itself allocates unconditionally, breaking several use cases that the C++ counterparts have - the naming is unfortunate and misleading.

Reply:
Just so that I understand you right: You are talking about custom deleters, right?

Reply:
custom deleters, passing in an existing pointer (that I got from some API), using with custom memory allocators / memory pools..

ie this type is a .. manager for allocshared really - as pointed out elsewhere, it's a nice manager, but still, the C++ type is quite a different beast for anyone that comes from C++.

on top of that, the quite rough in-place construction - C++, because it has constructors to begin with, has a more elegant solution in `make_unique` as others have pointed out - the way `newUniquePtr` returns an uninitialized memory area is really .. un-nim-like - for example, there's `newSeqUninitialized` which at least puts the unexpected behavior in the name.

Reply:
the same goes for its fit with nim itself: `ptr` and `pointer`  in nim are unsafe, unmanaged types - things where the GC is off - this type still behaves like a GC:d type because it manages the memory it allocates

Reply:
[Here](https://gist.github.com/planetis-m/b85ef61ae87160ddac802ce6df89d171) is the fusion.pools calling custom destructors. It's doable to return a ptr to a tuple (T, counter) instead.

Reply:
I see what you mean but I am not sure it's doable. Suppose we have `UniquePtr` that's a concept instead, there libs that provide their own implementation, one comes from a memory pool the other uses the std allocator...

Reply:
Just to be clear - I'm not necessarily of the opinion that the utilities in this module should have the same capabilities as the C++ equivalent - my main point is that the current naming fits neither Nim nor C++ based on what the module actually does - if anything, it behaves like a thread-sharable `UniqueRef`, whatever such a beast should be called.

Reply:
Take a look at https://github.com/nim-lang/threading/issues/7 please